File: systems/hover.lua
local system = require('library.system')
local viewport = require('utilities.viewport')

local mouse = love.mouse

return function(state)
    local s = system({'hover', 'position', 'appearance'})

    s.update = function(dt)
        local mouseX, mouseY = viewport.getMousePosition()
        local e
        for index = 1, #s.entities do
            e = s.entities[index]
            local width = e.appearance.width
            local height = e.appearance.height
            if e.animation then
                local _, _, w, h = e.animation.getFrame():getViewport()
                width = w
                height = h
            end
            local isWithinX = mouseX >= e.position.x and mouseX <= e.position.x + width
            local isWithinY = mouseY >= e.position.y and mouseY <= e.position.y + height
            local isHovered = isWithinX and isWithinY
            local hasHoverChanged = isHovered ~= e.hover.isHovered
            if hasHoverChanged then
                e.hover.isHovered = isHovered
                if e.animation and e.animation.sequences['hover'] then
                    if isHovered then
                        e.animation.previousSequence = e.animation.sequence
                        e.animation.sequence = 'hover'
                    else
                        e.animation.sequence = e.animation.previousSequence
                        e.animation.previousSequence = 'hover'
                    end
                end
            end
        end
    end

    return s
end

File: system/hover.go
package system

import (
	"github.com/leviceccato/clarity/game"

	"github.com/hajimehoshi/ebiten/v2"
)

type hoverSystem struct {
	System
}

func NewHoverSystem(s *game.State) *hoverSystem {
	sys := &hoverSystem{}
	sys.Components = []string{
		"Hover",
		"Position",
		"Size",
	}
	return sys
}

func (sys *hoverSystem) Load(s *game.State) {}

func (sys *hoverSystem) Update(s *game.State) {
	for _, e := range sys.Entities {
		// Check if mouse is within x and y ranges
		isHovered := s.CursorX >= e.Position.X && s.CursorX <= e.Position.X+e.Size.Width &&
			s.CursorY >= e.Position.Y && s.CursorY <= e.Position.Y+e.Size.Height

		// If no change, nothing to do
		hasHoverChanged := isHovered != e.Hover.IsHovered
		if !hasHoverChanged {
			return
		}

		e.Hover.IsHovered = isHovered
		s.SetIsCursorHovering(isHovered)

		// No animation, nothing to do
		if e.Appearance == nil {
			continue
		}

		// For animations allow toggling a 'hover' sequence if present
		_, hasHoverSequence := e.Appearance.Sequences["hover"]
		if !hasHoverSequence {
			continue
		}

		if isHovered {
			e.Appearance.PreviousSequence = e.Appearance.Sequence
			e.Appearance.Sequence = "hover"
			continue
		}
		e.Appearance.PreviousSequence = "hover"
		e.Appearance.Sequence = e.Appearance.PreviousSequence
	}
}

func (sys *hoverSystem) Draw(s *game.State, screen *ebiten.Image) {}

func (sys *hoverSystem) Enter(s *game.State) {}

func (sys *hoverSystem) Exit(s *game.State) {}
